### PR TITLE
skylog: Add Merkle tree build worker

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
This change introduces `BuildWorker` which is responsible for building lower
levels of the Merkle tree in accordance with leaf hashes that it gets.

### CODE REVIEW

File: skylog/core/builder.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	"context"
+	"fmt"
+
+	"github.com/google/trillian/merkle/compact"
+	"github.com/google/trillian/skylog/storage"
+)
+
+// BuildJob is a slice of leaf data hashes spanning a range of a Merkle tree.
+type BuildJob struct {
+	Begin  uint64   // The beginning of the range.
```
Comment (AlCutter) [inline skylog/core/builder.go]: How about `RangeStart`? (Begin sounds more like a function name)

File: skylog/core/builder_test.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	"reflect"
+	"testing"
+
+	"github.com/google/trillian/merkle/compact"
+	"github.com/google/trillian/merkle/rfc6962"
+	"github.com/google/trillian/merkle/testonly"
+	"github.com/google/trillian/skylog/storage"
+)
+
+// TODO(pavelkalinnikov): Use GoMock?
```
Comment (AlCutter) [inline skylog/core/builder_test.go]: Although the golang folks [used to?] hate on mocks, it does seem a little like you're having to reimplement some of that functionality here.

File: skylog/core/builder_test.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		return fmt.Errorf("got nodes: %+v\nwant: %+v", got, want)
+	}
+	return nil
+}
+
+func TestBuildWorker(t *testing.T) {
+	hashes := testonly.NodeHashes()
+	n := len(hashes[0])
+	if got, want := n, 8; got != want {
+		t.Fatalf("got %d leaves, want %d", got, want)
```
Comment (AlCutter) [inline skylog/core/builder_test.go]: nit: maybe change the logged message here - "prerequisite: testinly.NodeHashes() returned %d leaves, but this test wants 8" - just to make it clear that it's not the _test per se_ which has failed.

File: skylog/core/builder.go (hunk 1 above)
Comment (pavelkalinnikov) [reply skylog/core/builder.go]: Done.

File: skylog/core/builder_test.go (hunk 3 above)
Comment (pavelkalinnikov) [reply skylog/core/builder_test.go]: Done.

File: skylog/core/builder_test.go (hunk 2 above)
Comment (pavelkalinnikov) [reply skylog/core/builder_test.go]: Yep, made this a more definite TODO.

**Pull request merged**